### ISSUE TITLE
Wishlist: Katz centrality

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1977]
**What is the feature or improvement you would like to see?**

Implement Katz centrality. This requires some linear algebra, ideally done with sparse matrices for good performance. The implementation will likely need to use CXSparse.

**Use cases for the feature**

This may be the most popular centrality measure not yet implemented in igraph.

**References**

 - https://en.wikipedia.org/wiki/Katz_centrality

**Related**

 - `alpha_centrality()` in the R interface is related

**Comment by ntamas** [comment]
Are alpha centrality and Katz centrality the same thing under different names?

Judging from the code in the R interface, alpha centrality would require inverting a sparse matrix only. `cs_lsolve()` or one of the related functions would probably be enough.

**Comment by szhorvat** [comment]
> Are alpha centrality and Katz centrality the same thing under different names?

No, but they are closely related and it makes sense to implement them at the same time.



> Judging from the code in the R interface, alpha centrality would require inverting a sparse matrix only. `cs_lsolve()` or one of the related functions would probably be enough.

Yes, these are easy to implement in any language that has sparse linear algebra support.

**Comment by szhorvat** [comment]
> Are alpha centrality and Katz centrality the same thing under different names?

Indeed it seems that they are. The Wikipedia article is confusing and possibly inaccurate.

The centrality scores $x$ are defined by the equations

$$ x = \alpha A^T x + \beta, $$

where $\alpha$ is a scalar and $\beta$ is either a scalar, or a vector (a different value for each vertex, i.e. personalization). It seems that "Katz centrality" typically refers to the the scalar $\beta$ case and "alpha centrality" refers to the  personalized case.

----

The scores can be computed by solving the following linear system:

$$ (I - \alpha A^T) x = \beta.$$

Rescaling $\beta$ by a constant rescales the solution in the same way. Thus when $\beta$ is a scalar (no personalization), we might as well set it to 1, leaving $\alpha$ as the only parameter.

In order for this equation to posses a solution, we need $\det (I - \alpha A^T) \ne 0$, or equivalently $\det (A^T - \alpha^{-1} I)$ is not zero. This is simply the characteristic polynomial of $A$, thus it takes a zero value when $\alpha$ is the inverse of one of the eigenvalues of $A$. In fact as $\alpha^{-1}$ approaches an eigenvalue, $x$ will approach the corresponding eigenvector.

When $\alpha = 0$, $x = \beta = (1, 1, \dots)$.

The recommendation is thus to keep $0 \le \alpha < \frac{1}{\lambda_1}$, where $\lambda_1$ is the principal eigenvector.

When $\alpha > \frac{1}{\lambda_1}$, we run into the same problem as with eigenvector centrality: it is possible that not all scores will be non-negative.

----

What's missing for me before implementing this is clearer guidelines on choosing $\alpha$, especially in the personalized case. All other libraries I checked avoid trying to set $\alpha$ automatically, and leave it as the user's responsibility to choose an appropriate value.
